README.md
# DecentralTube Main API
[![Tests](https://github.com/MateoDi9z/express-template/actions/workflows/tests.yml/badge.svg)](https://github.com/MateoDi9z/express-template/actions/workflows/tests.yml)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DecentralTube/dct-main-api/master/LICENSE)

This project is meant to be the core API for DecentralTube, a 100% content-creator-owned (future blockchain based) video network.

## Features
- Express.js API 
- Content Provider
  - SQL DB (prisma)
- Unit tested (Mocha & Chai)
- Typescript

## Commands

Start development server
```sh
npm run start:dev
```

Run tests in /test dir
```sh
npm run test
```

Build project for production
```sh
npm run build
```

Start production server
```sh
npm run start:prod
```

[Git Repo](https://github.com/DecentralTube/decentraltube-api)

See [CHANGELOG](./CHANGELOG.md)

## License
This project is released under the [MIT License](https://opensource.org/licenses/MIT)

CHANGELOG.md
## What's Changed

## v1.0.3
* Changed file architecture by @MateoDi9z
* Create CRUD video API endpoint by @MateoDi9z
* Optimized old code & test by @MateoDi9z
* Edited Readme by @MateoDi9z
* Added @formats to module resolver by @MateoDi9z

## v1.0.2

* Prisma setup by @MateoDi9z
* Prisma User & Video Models by @MateoDi9z
* Create CRUD user API endpoint by @MateoDi9z
* Create CHANGELOG by @MateoDi9z
* Create `rest.http` to doc API endpoints by @MateoDi9z
* Add MIT license by @MateoDi9z

### v1.0.1

* Template init by @MateoDi9z

Go to [README](./README.md)

test/server.spec.ts
// import { expect } from "chai"
import request from "supertest"

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import createServer from "@src/server"
const app = createServer()

describe("> Server", () => {
  it("server is created without errors", (done) => {
    request(app).get("/").expect(200, done)
  })

  it("`/` responds with 'Hello World!'", (done) => {
    request(app).get("/").expect("Hello World!", done)

    //? SAME
    // request(app).get("/").then((res) => {
    //   expect(res.text).equal("Hello World!")
    //   done()
    // }).catch(e => {
    //   done(e)
    // })
  })
})


test/routes/user.spec.ts
import { expect } from "chai"
import request from "supertest"

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import createServer from "@src/server"
const app = createServer()

describe("> USER route", () => {
  //? Route check
  it("`/user` \t\t-> GET - returns OK", (done) => {
    request(app).get("/user").expect(200, done)
  })

  //? GET
  it("`/user` \t\t-> GET - responds with proper format", async () => {
    const response = await request(app)
      .get("/user")
      .then((res) => (res = res.body))
      .catch((e) => {
        throw new Error(e)
      })

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error")
    expect(response).to.have.own.property("status")
    expect(response).to.have.own.property("data")
  })

  //? Error messages
  it("`/user` \t\t-> GET - responds with good error messages", async () => {
    const response = await request(app)
      .get("/user")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    const data = response.data
    if (!data) {
      expect(response.error).to.be.true
      expect(response.status).equal(404 || 500)
      expect(response.data).to.be.null

      if (response.status == 404) expect(response.msg).equal("not found")
      if (response.status == 500) expect(response.msg).equal("server error")
    }
  })

  //? GET unique
  it("`/user/:id` \t-> GET - responds with one user", async () => {
    const created = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random()),
        username: String(Math.random())
      })
      .then((res) => res.body.data)
      .catch((e) => {
        throw new Error(e)
      })
    const response = await request(app)
      .get(`/user/${created.id}`)
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    await deleteTestUser(`/user/${response.data.id}`)

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error").to.be.a("boolean")
    expect(response).to.have.own.property("status")
    expect(response).to.have.own.property("data")
    if (response?.data) expect(response.data).to.be.an("object")
  })

  //? GET 404 Check
  it("`/user/1000` \t-> GET - responds with 404 error", async () => {
    const response = await request(app)
      .get("/user/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error").to.be.true
    expect(response).to.have.own.property("status").eql(404)
    expect(response).to.have.own.property("data").to.be.null
  })

  //? POST user
  it("`/user` \t\t-> POST - creates a user without errors", async () => {
    const created = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random()),
        username: String(Math.random())
      })
      .catch((e) => {
        throw new Error(e)
      })

    if (!created?.body) throw new Error("not created")

    expect(created.body).to.be.an("object")
    expect(created.body).to.have.own.property("error").to.be.false
    expect(created.body).to.have.own.property("status").to.eql(200)
    expect(created.body).to.have.own.property("data").to.be.an("object")

    const id = created.body.data.id
    const delete_url = `/user/${id}`
    await deleteTestUser(delete_url)
  })

  //? POST missing params check
  it("`/user` \t\t-> POST - throws error when missing params", async () => {
    const expectedError = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random())
      })
      .catch((e) => {
        throw new Error(e)
      })
    const error = expectedError.body
    expect(error).to.have.own.property("error").to.be.true
    expect(error).to.have.own.property("status").to.eql(400)
    expect(error).to.have.own.property("data").to.be.null
    expect(error).to.have.own.property("msg").to.eql("missing params")
  })

  //? DELETE user
  it("`/user/:id` \t-> DELETE - deletes user", async () => {
    const created = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random()),
        username: String(Math.random())
      })
      .catch((e) => {
        throw new Error(e)
      })

    if (created?.body) {
      const id = created.body.data.id
      const url = `/user/${id}`

      const deleted = await request(app)
        .delete(url)
        .then((res) => res.body)
        .catch((e) => {
          throw new Error(e)
        })

      expect(deleted).to.be.an("object")
      expect(deleted).to.have.own.property("error").to.be.false
      expect(deleted).to.have.own.property("status").to.eql(200)
      expect(deleted).to.have.own.property("data").to.be.an("object")
      expect(deleted.data.id).equal(id)
    } else throw new Error("not created")
  })

  //? DELETE 404 Check
  it("`/user/1000` \t-> DELETE - responds with 404", async () => {
    const deleted = await request(app)
      .delete("/user/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(deleted).to.be.an("object")
    expect(deleted).to.have.own.property("error").to.be.true
    expect(deleted).to.have.own.property("status").to.eql(404)
    expect(deleted).to.have.own.property("data").to.be.null
    expect(deleted).to.have.own.property("msg").to.eql("not found")
  })

  //? PUT user
  it("`/user/:id` \t-> PUT - edit user without errors", async () => {
    const created = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random()),
        username: String(Math.random())
      })
      .catch((e) => {
        throw new Error(e)
      })

    if (!created.body) {
      throw new Error("not created")
    }

    const created_id = created.body.data.id
    const edit_url = `/user/${created_id}`

    const edited = await request(app)
      .put(edit_url)
      .set("Content-Type", "application/json")
      .send({
        name: String(Math.random()),
        password: String(Math.random()),
        username: String(Math.random())
      })
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(edited).to.be.an("object")
    expect(edited).to.have.own.property("error").to.be.false
    expect(edited).to.have.own.property("status").to.eql(200)
    expect(edited).to.have.own.property("data").to.be.an("object")
    expect(edited.data.id).equal(created_id)

    if (!edited) {
      throw new Error("not edited")
    }

    const edited_id = edited.data.id
    const delete_url = `/user/${edited_id}`
    await deleteTestUser(delete_url)
  })

  //? PUT 404 Check
  it("`/user/1000` \t-> PUT - responds with 404", async () => {
    const update = await request(app)
      .put("/user/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(update).to.be.an("object")
    expect(update).to.have.own.property("error").to.be.true
    expect(update).to.have.own.property("status").to.eql(404)
    expect(update).to.have.own.property("data").to.be.null
    expect(update).to.have.own.property("msg").to.eql("not found")
  })
})

async function deleteTestUser(url: string) {
  await request(app)
    .delete(url)
    .catch((e) => {
      throw new Error(e)
    })
}


test/routes/video.spec.ts
import { expect } from "chai"
import request from "supertest"

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import createServer from "@src/server"
const app = createServer()

const route = "/" + "video"

let testUser: {
  id: number
  name: string
  password: string
  username: string
}

describe("> VIDEO route", () => {
  //* TEST USER
  before(async () => {
    const res = await request(app)
      .get("/user/username/test")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    if (res.data) {
      testUser = res.data
      return
    }

    testUser = await request(app)
      .post("/user")
      .set("Content-Type", "application/json")
      .send({
        name: "test",
        password: "test",
        username: "test"
      })
      .then((res) => res.body)
      .then((res) => res.data)
      .catch((e) => {
        throw new Error(e)
      })
  })

  //* Route check
  it(`"${route}" \t-> GET - returns OK`, (done) => {
    request(app).get(route).expect(200, done)
  })

  //* GET
  it(`"${route}" \t-> GET - responds with proper format`, async () => {
    const response = await request(app)
      .get(route)
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error")
    expect(response).to.have.own.property("status")
    expect(response).to.have.own.property("data")

    let data = response.data

    if (!data) {
      expect(data).to.be.null
      return
    }

    if (data[1]) {
      data = data[0]
      expect(response).to.have.own.property("count").to.be.a("number")
    }

    expect(data).to.have.own.property("id").to.be.a("number")
    expect(data).to.have.own.property("name").to.be.a("string")
    expect(data).to.have.own.property("password").to.be.a("string")
    expect(data).to.have.own.property("username").to.be.a("string")
    expect(data).to.have.own.property("createdAt").to.be.a("string")
    expect(data).to.have.own.property("updatedAt").to.be.a("string")
  })

  //* Error messages
  it(`"${route}" \t-> GET - responds with good error messages`, async () => {
    const response = await request(app)
      .get(route)
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    const data = response.data
    if (!data) {
      expect(response.error).to.be.true
      expect(response.status).equal(404 || 500)
      expect(response.data).to.be.null

      if (response.status == 404) expect(response.msg).equal("not found")
      if (response.status == 500) expect(response.msg).equal("server error")
    }
  })

  //* GET unique
  it(`"${route}/:id" \t-> GET - responds with one video`, async () => {
    const created = await request(app)
      .post(route)
      .set("Content-Type", "application/json")
      .send({
        title: "#1",
        video: String(Math.random()),
        authorUsername: testUser.username
      })
      .then((res) => res.body)
      .then((body) => body.data)
      .catch((e) => {
        throw new Error(e)
      })

    const response = await request(app)
      .get(`${route}/${created.id}`)
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    await deleteTestVideo(`${route}/${response.data.id}`)

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error").to.be.a("boolean")
    expect(response).to.have.own.property("status")
    expect(response).to.have.own.property("data")
    if (response?.data) expect(response.data).to.be.an("object")
  })

  //* GET 404 Check
  it(`"${route}/1000" \t-> GET - responds with 404 error`, async () => {
    const response = await request(app)
      .get(route + "/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(response).to.be.an("object")
    expect(response).to.have.own.property("error").to.be.true
    expect(response).to.have.own.property("status").eql(404)
    expect(response).to.have.own.property("data").to.be.null
  })

  //* POST video
  it(`"${route}" \t-> POST - creates a video without errors`, async () => {
    const created = await request(app)
      .post(route)
      .set("Content-Type", "application/json")
      .send({
        title: "#2",
        video: String(Math.random()),
        authorUsername: testUser.username
      })
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    if (!created) throw new Error("not created")

    expect(created).to.be.an("object")
    expect(created).to.have.own.property("error").to.be.false
    expect(created).to.have.own.property("status").to.eql(200)
    expect(created).to.have.own.property("data").to.be.an("object")

    const id = created.data.id
    const delete_url = `${route}/${id}`
    await deleteTestVideo(delete_url)
  })

  //* POST missing params check
  it(`"${route}" \t-> POST - throws error when missing params`, async () => {
    const expectedError = await request(app)
      .post(route)
      .set("Content-Type", "application/json")
      .send({
        title: "#3",
        video: String(Math.random())
      })
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(expectedError).to.have.own.property("error").to.be.true
    expect(expectedError).to.have.own.property("status").to.eql(400)
    expect(expectedError).to.have.own.property("data").to.be.null
    expect(expectedError).to.have.own.property("msg").to.eql("missing params")
  })

  //* DELETE video
  it(`"${route}/:id" \t-> DELETE - deletes video`, async () => {
    const created = await request(app)
      .post(route)
      .set("Content-Type", "application/json")
      .send({
        title: "#4",
        video: String(Math.random()),
        authorUsername: testUser.username
      })
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    if (created) {
      const id = created.data.id
      const url = `${route}/${id}`

      const deleted = await request(app)
        .delete(url)
        .then((res) => res.body)
        .catch((e) => {
          throw new Error(e)
        })

      expect(deleted).to.be.an("object")
      expect(deleted).to.have.own.property("error").to.be.false
      expect(deleted).to.have.own.property("status").to.eql(200)
      expect(deleted).to.have.own.property("data").to.be.an("object")
      expect(deleted.data.id).equal(id)
    } else throw new Error("not created")
  })

  //* DELETE 404 Check
  it(`"${route}/1000" \t-> DELETE - responds with 404`, async () => {
    const deleted = await request(app)
      .delete(route + "/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(deleted).to.be.an("object")
    expect(deleted).to.have.own.property("error").to.be.true
    expect(deleted).to.have.own.property("status").to.eql(404)
    expect(deleted).to.have.own.property("data").to.be.null
    expect(deleted).to.have.own.property("msg").to.eql("not found")
  })

  //* PUT video
  it(`"${route}/:id" \t-> PUT - edit video without errors`, async () => {
    const created = await request(app)
      .post(route)
      .set("Content-Type", "application/json")
      .send({
        title: "#5",
        video: String(Math.random()),
        authorUsername: testUser.username
      })
      .then((res) => res.body)
      .then((body) => body.data)
      .catch((e) => {
        throw new Error(e)
      })

    if (!created) {
      throw new Error("not created")
    }

    const edit_url = `${route}/${created.id}`

    const edited = await request(app)
      .put(edit_url)
      .set("Content-Type", "application/json")
      .send({
        title: "#6",
        video: String(Math.random()),
        authorUsername: testUser.username
      })
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(edited).to.be.an("object")
    expect(edited).to.have.own.property("error").to.be.false
    expect(edited).to.have.own.property("status").to.eql(200)
    expect(edited).to.have.own.property("data").to.be.an("object")
    expect(edited.data.id).equal(created.id)

    if (!edited) {
      throw new Error("not edited")
    }

    const edited_id = edited.data.id
    const delete_url = `${route}/${edited_id}`
    await deleteTestVideo(delete_url)
  })

  //* PUT 404 Check
  it(`"${route}/:1000" \t-> PUT - responds with 404`, async () => {
    const update = await request(app)
      .put(route + "/1000")
      .then((res) => res.body)
      .catch((e) => {
        throw new Error(e)
      })

    expect(update).to.be.an("object")
    expect(update).to.have.own.property("error").to.be.true
    expect(update).to.have.own.property("status").to.eql(404)
    expect(update).to.have.own.property("data").to.be.null
    expect(update).to.have.own.property("msg").to.eql("not found")
  })
})

async function deleteTestVideo(url: string) {
  await request(app)
    .delete(url)
    .catch((e) => {
      throw new Error(e)
    })
}


src/routes/user.ts
import "../moduleResolver"
import { Router, Request, Response } from "express"
import Format from "@class/responseFormat"

import { PrismaClient } from "@prisma/client"

const prisma = new PrismaClient()
const router = Router()

//* @route GET /user
//* @desc Get users
//* @access PRIVATE
router.get("/", async (_req: Request, res: Response) => {
  const data = await prisma.user.findMany()
  const resa = new Format(false, data)
  res.json(resa)
})

//* @route GET /user/:id
//* @desc Get user
//* @access PRIVATE
router.get("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)

  const data = await prisma.user.findUnique({
    where: {
      id: id
    }
  })

  res.json(data ? new Format(false, [data]) : new Format(true, null))
})

//* @route POST /user
//* @desc Post new user
//* @access PRIVATE
router.post("/", async (req: Request, res: Response) => {
  const { name, username, password } = req.body

  if (!name || !username || !password) {
    res.json(new Format(true, req.body, "missing params"))
  }

  const user = await prisma.user.create({
    data: {
      name,
      password,
      username
    }
  })

  res.json(new Format(false, [user]))
})

//* @route DELETE /user
//* @desc Delete user
//* @access PRIVATE
router.delete("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)

  const exists = await prisma.user.findUnique({
    where: {
      id: id
    }
  })

  if (!exists) {
    res.json(new Format(true, null))
  }

  const user = await prisma.user.delete({
    where: {
      id: id
    }
  })

  res.json(new Format(false, [user]))
})

//* @route PUT /user
//* @desc Update user
//* @access PRIVATE
router.put("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)
  const updatedUser = req.body

  const exists = await prisma.user.findUnique({
    where: {
      id: id
    }
  })

  if (!exists) {
    res.json(new Format(true, null))
  }

  const user = await prisma.user.update({
    where: {
      id: id
    },
    data: updatedUser
  })

  res.json(new Format(false, [user]))
})

export default router


src/routes/video.ts
import "../moduleResolver"
import { Router, Request, Response } from "express"
import Format from "@formats/video.format"

import { PrismaClient } from "@prisma/client"

const prisma = new PrismaClient()
const router = Router()

//* @route GET /video
//* @desc Get videos
//* @access PRIVATE
router.get("/", async (_req: Request, res: Response) => {
  const data = await prisma.video.findMany()
  res.json(new Format(false, data))
})

//* @route GET /video/:id
//* @desc Get video
//* @access PRIVATE
router.get("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)

  const data = await prisma.video.findUnique({
    where: {
      id: id
    }
  })

  res.json(data ? new Format(false, [data]) : new Format(true, null))
})

//* @route POST /video
//* @desc Post new video
//* @access PRIVATE
router.post("/", async (req: Request, res: Response) => {
  const { title, video, authorUsername } = req.body

  if (!title || !video || !authorUsername) {
    res.json(new Format(true, req.body, "missing params"))
    return
  }

  const newVideo = await prisma.video.create({
    data: {
      title,
      video,
      authorUsername
    }
  })

  res.json(new Format(false, [newVideo]))
})

//* @route DELETE /video
//* @desc Delete video
//* @access PRIVATE
router.delete("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)

  const exists = await prisma.video.findUnique({
    where: {
      id: id
    }
  })

  if (!exists) {
    res.json(new Format(true, null))
  }

  const video = await prisma.video.delete({
    where: {
      id: id
    }
  })

  res.json(new Format(false, [video]))
})

//* @route PUT /video
//* @desc Update video
//* @access PRIVATE
router.put("/:id", async (req: Request, res: Response) => {
  const id = parseInt(req.params.id)
  const updatedVideo = req.body

  const exists = await prisma.video.findUnique({
    where: {
      id: id
    }
  })

  if (!exists) {
    res.json(new Format(true, null))
  }

  const video = await prisma.video.update({
    where: {
      id: id
    },
    data: updatedVideo
  })

  res.json(new Format(false, [video]))
})

export default router


src/classes/formatters/user.format.ts
import { User } from "@prisma/client"

export default class Format {
  error: boolean
  msg: string
  status: number
  count?: number
  data: User[] | User | null

  constructor(error: boolean, data: User[] | null, msg?: string) {
    //* default values
    this.error = error
    this.status = error ? 500 : 200
    this.msg = error ? "server error" : "success"

    //* (400) - Missing param error
    if (msg == "missing params") {
      this.msg = msg
      this.status = 400
      this.data = null
      return
    }

    //* (404) - Not found error
    if (data == null || !data[0]) {
      this.error = true
      this.status = 404
      this.msg = "not found"
      this.data = null
      return
    }

    //* Format data
    if (data != null) {
      //* Multiple users
      if (data[1]) {
        this.count = data.length
        this.data = data
        return
      }
      //* One user
      this.data = data[0]
    } else this.data = null
  }
}


src/classes/formatters/video.format.ts
import { Video } from "@prisma/client"

export default class Format {
  error: boolean
  msg: string
  status: number
  count?: number
  data: Video[] | Video | null

  constructor(error: boolean, data: Video[] | null, msg?: string) {
    //* default values
    this.error = error
    this.status = error ? 500 : 200
    this.msg = error ? "server error" : "success"

    //* (400) - Missing param error
    if (msg == "missing params") {
      this.msg = msg
      this.status = 400
      this.data = null
      return
    }

    //* (404) - Not found error
    if (data == null || !data[0]) {
      this.error = true
      this.status = 404
      this.msg = "not found"
      this.data = null
      return
    }

    //* Format data
    if (data != null) {
      //* Multiple users
      if (data[1]) {
        this.count = data.length
        this.data = data
        return
      }
      //* One user
      this.data = data[0]
    } else this.data = null
  }
}
